### PR TITLE
Added float precision check in floor price comparision

### PR DESCRIPTION
**Pull request by pm-jaydeep-mohite**
bidadjustment factor: 0.9

original bidfloor: 4 USD

bidfloor in partner request 4/0.9 = 4.444444444444445

bid returned by partner : 4.44 USD

final bid price : 4.44-10% = 3.996 USD

in above scenario bid was getting rejected since bid price 3.99 is less than bidfloor 4 even when bidder returned bid which is equal to the bidfloor in request.

### CODE REVIEW

File: floors/enforce_test.go [hunk 1]
```diff
@@ -1154,7 +1219,66 @@ func TestUpdateEnforcePBS(t *testing.T) {
...
+	}
+}
+
+func TestIsBidPriceMoreThanBidfloor(t *testing.T) {
 
+	tests := []struct {
+		name      string
+		bidFloor  float64
+		bidPrice  float64
+		precision float64
```
Comment (hhhjort) [inline floors/enforce_test.go:1232]: The `precision` field does not seem to be used.

File: floors/enforce.go [hunk 2]
```diff
@@ -154,6 +155,14 @@ func enforceFloorToBids(bidRequestWrapper *openrtb_ext.RequestWrapper, seatBids
 	return seatBids, errs, rejectedBids
 }
 
+// isBidPriceLowerThanBidfloor checks for bid price with bid floor with given precision
+func isBidPriceLowerThanBidfloor(bidFloor float64, bidPrice float64) bool {
+	if bidPrice < bidFloor && math.Abs(bidFloor-bidPrice) > floorPrecision {
```
Comment (hhhjort) [inline floors/enforce.go:160]: Would this not be the same as `(bidPrice + floorPrecision) < bidFloor`

**Comment by hhhjort** [comment]
In your example, the bid floor is 4.444444444444445. If you round to the nearest cent, you round down, so 4.44 is technically less than 4.444444444444445. The bidder would have to bid 4.45 in order to win in the old code, which is the lowest whole cent value that is at or above the given bid floor.

In practice, I don't see this mattering unless the bidder knows both the exact bid floor and the bid adjustment value applied to them. Knowing the bid floor is common, but I thought publishers typically wish to keep the bid adjustment values private from the bidders. Is it common for bidders to know the bid adjustment value that is getting applied to them?

File: floors/enforce_test.go (hunk 1 above)
Comment (pm-jaydeep-mohite) [reply floors/enforce_test.go:1232]: Removed unused variable

File: floors/enforce.go (hunk 2 above)
Comment (pm-jaydeep-mohite) [reply floors/enforce.go:160]: corrected

**Comment by pm-jaydeep-mohite** [comment]
> In your example, the bid floor is 4.444444444444445. If you round to the nearest cent, you round down, so 4.44 is technically less than 4.444444444444445. The bidder would have to bid 4.45 in order to win in the old code, which is the lowest whole cent value that is at or above the given bid floor.
> 
> In practice, I don't see this mattering unless the bidder knows both the exact bid floor and the bid adjustment value applied to them. Knowing the bid floor is common, but I thought publishers typically wish to keep the bid adjustment values private from the bidders. Is it common for bidders to know the bid adjustment value that is getting applied to them?

Precision margin is required when precision loss occurred due to bidAdjustment for any bidder is present, otherwise no need of precision margin.

**Review by hhhjort — APPROVED** [review]
LGTM

**Review by AlexBVolcy — APPROVED** [review]

**Pull request merged**